accord-core/src/main/java/accord/coordinate/CoordinateEphemeralRead.java

```diff
@@ -0,0 +1,124 @@
+
+import static accord.coordinate.tracking.RequestStatus.Failed;
+import static accord.coordinate.tracking.RequestStatus.Success;
+
+public class CoordinateEphemeralRead extends AbstractCoordinatePreAccept<Result, GetEphemeralReadDepsOk>
+{
+    public static AsyncResult<Result> coordinate(Node node, FullRoute<?> route, TxnId txnId, Txn txn)
+    {
+        TopologyMismatch mismatch = TopologyMismatch.checkForMismatch(node.topology().globalForEpoch(txnId.epoch()), txnId, route.homeKey(), txn.keys());
+        if (mismatch != null)
+            return AsyncResults.failure(mismatch);
+
+        Topologies topologies = node.topology().withUnsyncedEpochs(route, txnId, txnId);
+        CoordinateEphemeralRead coordinate = new CoordinateEphemeralRead(node, topologies, route, txnId, txn);
+        coordinate.start();
+        return coordinate;
+    }
+
+    private final QuorumTracker tracker;
+    private final List<GetEphemeralReadDepsOk> oks;
+    private long executeAtEpoch;
+
+    CoordinateEphemeralRead(Node node, Topologies topologies, FullRoute<?> route, TxnId txnId, Txn txn)
+    {
+        super(node, route, txnId, txn);
+        this.tracker = new QuorumTracker(topologies);
+        this.executeAtEpoch = txnId.epoch();
+        this.oks = new ArrayList<>(topologies.estimateUniqueNodes());
+    }
+
+    @Override
+    void contact(Set<Node.Id> nodes, Topologies topologies, Callback<GetEphemeralReadDepsOk> callback)
+    {
+        CommandStore commandStore = CommandStore.maybeCurrent();
+        if (commandStore == null) commandStore = node.commandStores().select(route.homeKey());
+        node.send(nodes, to -> new GetEphemeralReadDeps(to, topologies, route, txnId, txn.keys(), executeAtEpoch), commandStore, callback);
+    }
+
+    @Override
+    long executeAtEpoch()
+    {
+        return executeAtEpoch;
+    }
+
+    @Override
+    public void onSuccessInternal(Node.Id from, GetEphemeralReadDepsOk ok)
+    {
+        oks.add(ok);
+        if (ok.latestEpoch > executeAtEpoch)
+            executeAtEpoch = ok.latestEpoch;
+
+        if (tracker.recordSuccess(from) == Success)
+            onPreAcceptedOrNewEpoch();
+    }
+
+    @Override
+    boolean onExtraSuccessInternal(Node.Id from, GetEphemeralReadDepsOk ok)
+    {
+        if (ok.latestEpoch > executeAtEpoch)
+            executeAtEpoch = ok.latestEpoch;
+
+        oks.add(ok);
+        return true;
+    }
+
+    @Override
+    public void onFailureInternal(Node.Id from, Throwable failure)
+    {
+        if (tracker.recordFailure(from) == Failed)
+            setFailure(new Timeout(txnId, route.homeKey()));
+    }
+
+    @Override
+    void onNewEpochTopologyMismatch(TopologyMismatch mismatch)
+    {
+        accept(null, mismatch);
+    }
+
+    @Override
+    void onPreAccepted(Topologies topologies)
+    {
+        Deps deps = Deps.merge(oks, ok -> ok.deps);
+        new ExecuteEphemeralRead(node, topologies, route, txnId, txn, executeAtEpoch, deps, this).start();
```

Comment:
Can we bump the Epoch in the `txnid` for the read to the `executeAtEpoch` so that when the read happens it can check if the epoch has a migration happening?

I think that would address the issue of doing a paxos write with async commit, then doing an Accord read at a different node that doesn't know about migrating to Paxos and having Accord accept the read. With the later Epoch Cassandra will reject the read in `TxnQuery`.

I think this gives you read your own writes in this base case. There might be more compleax cases where it still doesn't hold because the read can still occur in a later epoch, but it at least puts in a floor.

Reply:
Hmm, maybe what we want then is to use `TxnId.withEpoch(executeAtEpoch)` as the `executeAt`?

Reply:
Could you point me to where this check happens?

Reply:
It's here https://github.com/apache/cassandra/blob/00172b5fa9a935897a30b918964c2eaa4ccfd556/src/java/org/apache/cassandra/service/accord/txn/TxnQuery.java#L155

Thinking through this again. An ephemeral read needs to not read something already migrated back to Paxos, or read something not yet migrated to Accord. In Paxos -> Accord I think this works because there are no ephemeral reads so the interop checks at prepare/propose should catch it. 

### Accord -> Paxos 
In Accord -> Paxos a coordinator that doesn't know about the migration can create an ephemeral read and get it accepted because Accord doesn't know about migration only the integration does and the integration only checks when the read actually tries to run. The current approach just assumes that all keys are no longer readable/writable in the migration epoch regardless of whether the key has been migrated.

Because ephemeral reads aren't visible to other transactions (such as the barrier migrating the key to Paxos) the ephemeral read could run after the key is migrated to Paxos. It could run at any time really so it could end up executing in a later epoch even if the `executeAt` is in an earlier epoch.

However since it is a single key and single replica read maybe we can validate that the ephemeral read is safe in the integration by checking if a Paxos write occurred? It would need to block other Paxos writes while it reads the key. This is basically substituting for the ephemeral read not blocking other txns inside Accord.

Reply:
To close the loop on this. Ephemeral reads *do* execute in their `executeAt` epoch, it's validated according to Benedict. I haven't found where yet though. So that means the code in `TxnQuery` that validates things will be able to work.